docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,113 +96,233 @@ The current multisig public keys use amino serialization to generate the address
+
+Implementation Tip: account implementations should cache address in their structure.
 
-### Multisig Addresses
+#### Multisig Addresses
 
-For new multisig public keys, we define a custom address format not based on any encoding scheme
-(amino or protobuf). This avoids issues with non-determinism in the encoding scheme. It also
-ensures that multisig public keys which differ simply in the ordering of keys have the same
-address by sorting child public keys first.
+For new multisig public keys, we define the `typ` parameter not based on any encoding scheme (amino or protobuf). This avoids issues with non-determinism in the encoding scheme.
+
+Example:
 
-First we define a proto message for multisig public keys:
 ```proto
 package cosmos.crypto.multisig;
 
 message PubKey {
   uint32 threshold = 1;
-  repeated google.protobuf.Any public_keys = 2;
+  repeated google.protobuf.Any pubkeys = 2;
 }
 ```
 
-We define the following `Address()` function for this public key:
-
-```
+```go
 func (multisig PubKey) Address() {
-  // first gather all the addresses of each nested public key
-  var addresses [][]byte
-  for key := range multisig.Keys {
-    addresses = append(joinedAddresses, key.Address())
+  // first gather all nested pub keys
+  var keys []address.Addressable  // cryptotypes.PubKey implements Addressable
+  for _, _key := range multisig.Pubkeys {
+    keys = append(keys, key.GetCachedValue().(cryptotypes.PubKey))
   }
 
-  // then sort them in ascending order
-  addresses = Sort(addresses)
+  // form the type from the message name (cosmos.crypto.multisig.PubKey) and the threshold joined together
+  prefix := fmt.Sprintf("%s/%d", proto.MessageName(multisig), multisig.Threshold)
 
-  // then concatenate them together
-  var joinedAddresses []byte
-  for addr := range addresses {
-    joinedAddresses := append(joinedAddresses, addr...)
-  }
+  // use the Composed function defined above
+  return address.NewComposed(prefix, keys)
+}
+```
 
-  // form the string prefix from the message name (cosmos.crypto.multisig.PubKey) and the threshold joined together
-  prefix := fmt.Sprintf("%s/%d", proto.MessageName(multisig), multisig.Threshold)
+#### Module Account Addresses
 
-  // use the standard AddressHash function
-  return AddressHash(prefix, joinedAddresses)
+NOTE: this section is not finalize and it's in active discussion.
+
+In Basic Address section we defined a module account address as:
+
+```
+address.Hash("module", moduleName)
+```
+
+We use `"module"` as a schema type for all module derived addresses. Module accounts can have sub accounts. The derivation process has a defined order: module name, submodule, key, subsubmodule key.
+Module account addresses are heavily used in the SDK so it make sense to optimize the derivation process: instead of using of using `LengthPrefix` for the module name, we use a null byte (`'\x00'`) as a separator. This works, because null byte is not a part of a valid module name.
```

Comment:
```suggestion
Module account addresses are heavily used in the SDK so it makes sense to optimize the derivation process: instead of using of using `LengthPrefix` for the module name, we use a null byte (`'\x00'`) as a separator. This works, because null byte is not a part of a valid module name.
```